### PR TITLE
Switch to Towncrier

### LINKED ISSUE #10334: Use a YAML file as the basis for the change log
### Description

The repo at https://github.com/taldcroft/astropy-changelog shows a proof of concept for changing the basis of the astropy changelog from the `CHANGES.rst` file to a new `CHANGES.yml` file.  Please see the README there for details.  The driver for this and initial discussions were captured in an astropy-dev email thread in Oct 2016 (subject "Proposal for a new changelog system"). 

This was discussed in more detail at the 2019 coordination meeting in Socorro, with the agreement at the time that @bsipocz and @Cadair would develop a different system that uses information embedded in the PR description and/or GitHub metadata. While arguably a better solution, 6 months have passed with no progress, and in fact the details of the better solution were not captured in any written form. @bsipocz has stated in private conversation that she does not have time to spend on this given many other higher priorities and is OK with the YAML solution.

I propose moving forward on the YAML solution unless @Cadair can commit to developing the better solution in the next few months. All details of the YAML solution and how it gets implemented (both code and process) are certainly open for discussion here.

The problems in the current CHANGES.rst implementation are real and persist. I personally had to rebase at least 5 times due to a changelog conflict in the last few weeks (including once literally ten minutes after the last rebase), and now I see the 5-day old #10296 is already blocked from merge because of a changelog conflict. I also made a mistake in the placement of a change entry. 

### Additional context

https://github.com/taldcroft/astropy-changelog

### PR DESCRIPTION
**Pull request by saimn**
Ref #10334 : this is to show that we can use towncrier and it supports all the features we need.
Change fragments would go to `docs/changes/{subpackage}/` (can be changed) and I added a `docs/changes/README.rst` based on the Numpy one to explain how to add a changelog entry.
Then there is a template (`docs/changes/template.rst`) to get the same format as now, and the config with sections and entry types is done in `pyproject.toml`.
To test on a real case I converted the changes we have for 4.3 to change fragments (thanks to @taldcroft's parsing script from #10334). Running `towncrier build --draft --version 4.3` gives this: https://gist.github.com/saimn/36cca0290240b2159cece0dfdd243d22

EDIT: Close #10334

### CODE REVIEW

**Comment by saimn** [comment]
Also the main change that should be done if we go this way is to split the changelog in one changelog per version. (And where should we put those ?)

**Comment by taldcroft** [comment]
@saimn - thanks for this proof of concept! This helps a lot and I'm happy enough to consider going this way.

I like the sub-sections as permanent directories. That serves a good purpose of defining available subsections.

The concern about not being able to see the rendered version of the change log in master (on GitHub) still applies, i.e. there is not any easy way to see a summary of all the changes that have been committed to the release. One option is to just accept that and say you need to go into a terminal and use towncrier to generate the updates. But as a typical user this is a bit tricky because you need to grab the milestone from GitHub for the summary to be accurate. I do frequently look at the recent changes in the changelog, so we don't want to give up on that capability too quickly (though of course other projects seem to manage).

Another smaller issue is the case of multiple PRs that belong to one change. This happens sometimes for very big changes where you have the main PR and then one or more follow-on PR's to address known issues that came up in review. I can also happen with bugs in a PR that were fixed in dev before release. Will there be any way to allow this?

**Comment by taldcroft** [comment]
One more concern to discuss is the migration from a single entire-history changelog to per-version changelogs. Since the community has come to expect the entire-history changelog and may have links to this file in different places, we have to be aware of breakage like that.

There is also the issue that finding the version associated with some change by a text search in CHANGES.rst is no longer easy. Not a showstopper but still a change to consider.

**Comment by Cadair** [comment]
> The concern about not being able to see the rendered version of the change log in master (on GitHub) still applies

While not perfect, SunPy solves this with a [sphinx extension](https://github.com/sunpy/sunpy/blob/master/sunpy/util/sphinx/changelog.py) which renders it to our documentation: https://docs.sunpy.org/en/latest/whatsnew/changelog.html which shows you everything that's in the log since the last release.

**Comment by taldcroft** [comment]
And to be clear, I'm just applying a high bar for making a change like this in my comments, but I'm generally :+1: on this and very happy that @saimn took the initiative to actually show this can work for us!

**Comment by Cadair** [comment]
> There is also the issue that finding the version associated with some change by a text search in CHANGES.rst is no longer easy. Not a showstopper but still a change to consider.

This is true once the release is done and the final log rendered? or am I missing something?

**Comment by taldcroft** [comment]
> This is true once the release is done and the final log rendered? or am I missing something?

I got the impression there will be only per-version changelogs (not a single comprehensive one), so if you don't know which version the change was in then you have to search through files. But I might not be understanding.

**Comment by Cadair** [comment]
I haven't looked through the changes in this PR in detail, but sunpy still renders out to a single file on release: https://github.com/sunpy/sunpy/blob/master/CHANGELOG.rst

**Comment by Cadair** [comment]
> But just looking at the diff here, there is no indication of milestone anywhere and yet it magically know all these files are for 4.3; how

You assign the release to the entries at render time, this is the key component of towncrier to me. An entry has no version associated with it until that version is released with the fragement included. This means there is no issuies with backporting changelog entries they come along for the ride and then are assigned to the version as part of the release procedure.

> For us to update the change log bot to check that change log needs to be there or is correct?

Baldrick/Giles supports verifying the presence of a towncrier changelog entry and that the number matches the PR and that the type is one of the allowed types.

**Comment by pllim** [comment]
> You assign the release to the entries at render time

Can you please clarify -- So the release manager renders it, and then put the rendered file under version control?

**Comment by pllim** [comment]
This also means, we need to implement this solution for 4.0.x and 4.2.x branches? Otherwise, the release manager needs to handle two different change log mechanisms.

**Comment by saimn** [comment]
Great to see that you're happy with this @taldcroft :)

> The concern about not being able to see the rendered version of the change log in master (on GitHub) still applies

The Sphinx extension seems a good solution. Would be great if you can move it to a dedicated package @Cadair (also I don't understand your comment "Which I now see that @saimn has done something similar to here". There is no sphinx extension in my PR ?)

> Another smaller issue is the case of multiple PRs that belong to one change.

It is possible to have duplicated fragments with different PR number or sections/types. For example when converting the existing changs for 4.3 to fragments, some entries have two PR/issue numbers, which I converted to two fragment files: 
```
❯ cat docs/changes/io.fits/11149.feature.rst
Added ``append`` keyword to append table objects to an existing FITS file
❯ cat docs/changes/io.fits/2632.feature.rst 
Added ``append`` keyword to append table objects to an existing FITS file
```
And towncrier correctly renders this :
```
- Added ``append`` keyword to append table objects to an existing FITS file [#2632, #11149]
```

> I haven't looked through the changes in this PR in detail, but sunpy still renders out to a single file on release: https://github.com/sunpy/sunpy/blob/master/CHANGELOG.rst

Great, I did not know that was possible!

**Comment by saimn** [comment]
@pllim - About milestones, that's the killer feature with towncrier, there is no need to deal with it. Each PR must have a "changelog fragment", for example:
```
❯ cat docs/changes/io.fits/11149.feature.rst
Added ``append`` keyword to append table objects to an existing FITS file
```
Then the PR can be backported with its fragment, or included in the release branch. And when doing the release the changelog is rendered (with the `towncrier build` command) and all the fragment in `docs/changes/` are included. And then the fragments are deleted (`towncrier build` can delete them). So this way each release branch contains the fragments for the PRs that are included in the branch.

> Also, has towncrier ever introduce breaking change that needs downstream to refactor entries?

I don't think so. The content of the file is just plain RST that will go in the final changelog file. What could change a bit is the syntax for the filename, but the changes I've seen are mostly adding flexibility with more options to customize it.

**Comment by Cadair** [comment]
> I don't understand your comment "Which I now see that @saimn has done something similar to here"

oh I think I looked at the changelog template and thought it was a docs thing because of the path.

> Would be great if you can move it to a dedicated package

I am happy to put the effort into doing this, but only after this PR gets enough approval behind it that I am confident it will actually be needed ;)

> Can you please clarify -- So the release manager renders it, and then put the rendered file under version control?

The flow for sunpy is as follows:

* PRs come with one or more associated fragment files.
* The feature branch is made, all of the fragment files on the main branch are removed.
* As part of the final release the command `towncrier --version vX.Y.Z` is run which renders the changelog to a new section at the front of our `CHANGELOG.rst` file.
* All the fragement files are deleted (as part of the above command)
* The rendered changelog and the removal of the fragments are committed and released.

Any backports have their fragment files backported with the code changes, and these fragment files are deleted from the release branches at the time of release.

> This also means, we need to implement this solution for 4.0.x and 4.2.x branches? Otherwise, the release manager needs to handle two different change log mechanisms.

If a bugfix release of those branches was made aligned with the transition all future release of those branches would be able to just start using the new flow without any further changes (i.e. towncrier becomes the WIP changelog section at the top of the file for those branches after that release). If that didn't happen any WIP changelog entries would need converting to fragments.

**Comment by saimn** [comment]
> This also means, we need to implement this solution for 4.0.x and 4.2.x branches?

Not sure what is best for the transition. I guess we could backport this change after having backported all the PRs that need to and used the old way. Then with the little script I have we can convert the un-released changes to fragments, for each release branch.

**Comment by pllim** [comment]
> these fragment files are deleted from the release branches at the time of release

Just the release branch or also `master` branch?

**Comment by saimn** [comment]
I think the best would be to cherry-pick the commit rendering the changelog and deleting the fragments to master, for each release branch, as we want the fragments to be deleted for the next version and the changelog to include all the released versions.

**Comment by pllim** [comment]
All this adding and deleting possibly hundreds of files each release won't be a problem for the repo size? (I guess not but I want to confirm.)

**Comment by taldcroft** [comment]
All my concerns have been addressed, fantastic! I'm pleasantly surprised that towncrier is clever enough to recognize that fragments are identical and make just one entry.

So :+1: from me on this.

I think we need to get concurrence from @pllim and at least one of the release managers @bsipocz @eteq @astrofrog for provisional approval using towncrier (so e.g. @cadair can invest time).

**Comment by pllim** [comment]
I don't see any real blocker but who is doing the 4.3 release? That person needs to be on board since this person will be handling all the towncrier fragments and so on.

**Comment by bsipocz** [comment]
With the sphinx rendering to the dev version of the docs my main concern is addressed (as it also addresses the milestone and conflict issues without going into formatting requirements from contributors). It's still not as convenient as the current singe, conveniently human-readable single rst file, but I can live with it.
 One still fuzzy bit is about the bugfix releases. E.g. will stuff go into 4.2.1 end up in 4.3, too or not? I still think it should not, after all those releases are considered the latest stable at their point of release. 
I'm a bit hesitant not to duplicate entries from LTS either, but there I'm more flexible to accept other ideas.

One other minor point: at release time there was often a need to minor formatting, language, add or remove type of adjustments to the changelog. Should one need to go into the individual files, or it's possible to edit the rendered version before committing it to the release?

**Comment by Cadair** [comment]
>  Should one need to go into the individual files, or it's possible to edit the rendered version before committing it to the release?

You can render the changelog and then edit it at release time.

> One still fuzzy bit is about the bugfix releases. E.g. will stuff go into 4.2.1 end up in 4.3, too or not? I still think it should not, after all those releases are considered the latest stable at their point of release.

A patch which gets backported ends up in the log for both versions. This is pretty unavoidable given the way towncrier works.

**Comment by bsipocz** [comment]
>>    One still fuzzy bit is about the bugfix releases. E.g. will stuff go into 4.2.1 end up in 4.3, too or not? I still think it should not, after all those releases are considered the latest stable at their point of release.

> A patch which gets backported ends up in the log for both versions. This is pretty unavoidable given the way towncrier works.

Duplicating entries is still a big 👎 from me, but if we get a solution/workflow for that (e.g. deleting the files from master once the bugfix release is out), I would be happy with this approach.

**Comment by Cadair** [comment]
To me it is a minor change in behaviour of the changelog. Having the 4.3.0 changelog be the list of changes since 4.2.0 rather than 4.2.x doesn't seem unreasonable, it's just different. 

If the people doing the releases want to delete the files from master once the release is done then that's obviously fine though, the easiest way to do that would be to have a single commit which is just the changes made by running the towncrier command and "forwardport" that to master. That would mean that a) the master `CHANGES.rst` file has all the bug fix releases in it and b) the fragments get deleted from master after the bug fix.

Where this plan gets messy is that changes on the LTS branch and the bugfix branch will conflict on the deletion for the one which is forwardported last.

**Comment by saimn** [comment]
For me duplicating entries is a great feature ;). But @bsipocz, we discussed that already in the other PR and I thought you were finally happy with it (https://github.com/astropy/astropy/issues/10334#issuecomment-690621633).
As explained there, I think a given change should be mentioned in each version it is released. Because otherwise it's very difficult to know which bugs were fixed in e.g. 4.3. You have to know which bugfix release is released exactly at the same time and to know that bugfixes listed there are also in the new major release. 
Also this puts a constraint on releasing minor and major versions at the same time, which could be avoided and could avoid extra delays because a release is waiting on another one.

**Comment by taldcroft** [comment]
> Duplicating entries is still a big 👎 from me

I consider this a feature not a bug.  If I'm upgrading to 4.3.0 then I'm not going to be looking at changes in 4.2.x for bug fixes. I would like to see every PR that has been put into the 4.3.0 release.

**Comment by astrofrog** [comment]
I agree that the duplicate entries are good, as otherwise it does indeed force us to always release a minor release immediately prior to every major release.

**Comment by bsipocz** [comment]
@saimn - duplicating between lts and stable is fine with me, but duplicating between bugfix and next feature is a mistake I think. With that logic we could duplicate everything forever, after all there could be people who upgrade from 3.1, they would only interested in what changed compared to their version and not of what changed since the last release

**Comment by bsipocz** [comment]
And I agree of notut doing bugfix release at the time of feature release. We still need to do an lts (I think timing those to be around stable releases is good), but there isn't much point in doing a bugfix when we're dropping support for that branch. 

With the new changlog system this will be easy to not do, previously it required a lot of remilestoning and manual shuffling of entries

**Comment by saimn** [comment]
Oh: https://github.com/sphinx-contrib/sphinxcontrib-towncrier
Did you know about this one @Cadair ?

**Comment by Cadair** [comment]
No I did not, that certainly looks more complex than my hack... want to try it out here?

**Comment by saimn** [comment]
Yes, will try. Better if we can avoid making a new sphinx puppy.

**Comment by Cadair** [comment]
https://github.com/sphinx-contrib/sphinxcontrib-towncrier/issues/44 isn't encouraging.

**Comment by embray** [comment]
I haven't looked into this closely, but if @taldcroft (whose custom changelog solution I liked) is convinced, then I'm :+1: as well (having also supported towncrier in the past).

**Comment by pllim** [comment]
I feel like we are really close here but need the people who still disagree on some implementation details (particularly the existence of duplicate entries) to meet and hash things out. See https://groups.google.com/g/astropy-core-maintainers/c/37eIwivefeo

**Comment by pllim** [comment]
Decisions from 2021-02-25 meeting below. cc @saimn @Cadair @bsipocz @astrofrog @eteq 

## Accepted solution to handle duplicate entries

Forward port from maintenance branches to master *until* the LTS break point, from then on LTS branch is completely separated and has an independent changelog and nothing is forward-ported from LTS to master (And for non-LTS maintenance branches, no maintenance required after next major release)

p.s. SunPy's workflow was also discussed but we decided for that to be future work.

## Sphinx extension

We definitely want changelog rendered on `master`. We will pick the solution that works the way we want and is least work/maintenance. If the effort is low, we can include in this PR. If not, it needs to go in ASAP after this is merged to minimize the period we lack rendered human-readable changelog on `master`.

## Bot

@Cadair has a clear plan on how to disable `astropy-bot` and enable `baldrick` to go with this PR. He even stated that he could push that change straight to this PR. He and @saimn could coordinate on this.

## Path forward

This is laid out in #11351

**Comment by saimn** [comment]
I updated the various documentation pages mentioning the changelog, hopefully I did not miss some. The releasing doc will certainly need more updates, but I guess that's something to do when doing the first release with towncrier.

@Cadair : I had a look at the other Sphinx extension, which achieved to convince me of using the Sunpy one ;-). If you can move it to an independent package when you find the time, would be great !

**Comment by Cadair** [comment]
I have done the initial release of the changelog sphinx package: https://github.com/OpenAstronomy/sphinx-changelog & https://pypi.org/project/sphinx-changelog/

I will tag a non-prerelease when I can get Azure to auto-upload the dists, and once we know it works here.

**Comment by Cadair** [comment]
Oh I will push a commit or two here soon with the changes for the extension and the bot.

**Comment by pllim** [comment]
RTD timed out at 2000 s. Is towncrier supposed to take that long?

**Comment by Cadair** [comment]
I am waiting for pypi to unbreak itself so I can upload a fixed version of the extension. I have a suspicion my extension isn't parallel safe though, hence the timeout.

**Comment by Cadair** [comment]
It works :tada: https://astropy--11279.org.readthedocs.build/en/11279/changelog.html

File: docs/development/releasing.rst [hunk 1]
```diff
@@ -92,15 +92,14 @@ packages that use the full bugfix/maintenance branch approach.)
...
-   the title of the changelog section should be replaced too, thus getting rid
-   of any mention of the release candidate.
-   Then add and commit those changes with::
-
-      <use your favorite editor on CHANGES.rst>
+#. Render the changelog with towncrier, and confirm that the fragments can be
+   deleted. (Note: update this when doing the next release!)
+   ::
+
+       towncrier build [--draft] --version 4.3
```
Comment (Cadair) [inline docs/development/releasing.rst:99]: this only works on the master branch.

```suggestion
       towncrier [--draft] --version 4.3
```

File: pyproject.toml [hunk 2]
```diff
@@ -8,7 +8,154 @@ requires = ["setuptools",
             "extension-helpers"]
```
Comment (Cadair) [inline pyproject.toml:8]: The indentation in this file is all horribly inconsistent and some of it is my fault. Someone needs to crack it in an editor and pick an indentation pattern.

**Review by Cadair — COMMENTED** [review]
I am not sure there is much else to do here now?

File: pyproject.toml (hunk 2 above)
Comment (pllim) [reply pyproject.toml:8]: I reformatted it on this branch. Please double check.

**Comment by pllim** [comment]
Thanks! I guess we won't see if the bot really works or not until after merge?

**Comment by Cadair** [comment]
> Thanks! I guess we won't see if the bot really works or not until after merge?

yeah baldrick only reads HEAD config.

File: docs/development/releasing.rst (hunk 1 above)
Comment (Cadair) [reply docs/development/releasing.rst:99]: @saimn should we commit this?

Comment (saimn) [reply docs/development/releasing.rst]: @Cadair - Sure, thanks!

**Comment by saimn** [comment]
I rebased on master and convert the 4.3 changelog to towncrier fragments, so I think this is ready !

**Comment by Cadair** [comment]
4.2.1 hasn't made it out into the world yet right?

**Review by Cadair — APPROVED** [review]
I don't think this needs to be held on the release of 4.2.1 a followup PR can made any adjustments (as would normally happen) after that release?

**Comment by saimn** [comment]
Indeed 4.2.1 is not out yet but this is not a problem (since I don't think there will be new PRs on master that would be backported to 4.2.1). And it would be good to merge this PR before merging PRs on master that will need to be backported for future bugfix releases.

**Comment by pllim** [comment]
I don't think there is any immediate rush. Does this really need to go in *right now* as opposed to waiting for the release, which will hopefully happen this week?

**Comment by saimn** [comment]
No rush but if PRs are merged in master with the old changelog and must be backported, then some additional work will be needed when doing the bugfix releases. So currently I'm avoiding to merge PRs. Also If new PRs get merged I need to update this PR to create the fragments.

**Comment by pllim** [comment]
Ah, okay... I wonder if there is a way to "lock" `master` until this gets in. 💭

**Comment by pllim** [comment]
I temporarily enforced `At least 6 approving reviews are required by reviewers with write access.` So maybe that would block other PRs from getting merged until this is in for `master`.

Merges into `v4.0.x` and `v4.2.x` do not affect your work here, right?

**Comment by pllim** [comment]
4.0.5 and 4.2.1 are released. Should we wrap this up and merge? @saimn , do you want to do one last sweep? I am also not sure how this is compatible with #11461 .

**Review by Cadair — APPROVED** [review]
LET'S DO THIS

**Review by pllim — APPROVED** [review]
I guess we can merge now and figure out if the bot and the doc works after the fact?

Thanks for all the work here!

**Comment by pllim** [comment]
#yolo

**Comment by taldcroft** [comment]
🚀  !!  I think it would be good to follow up @pllim 's email to astropy-dev on March 1 with another saying this has now happened.

**Comment by pllim** [comment]
To close the loop, @saimn posted this follow up on the mailing list. Thank you, all!

<hr/>

Hi everyone,

The pull request has been merged so astropy is now using towncrier [1] to manage its changelog. If you have a pull request open with the old changelog system you must update it to the new system:

- rebase on master
- create the fragment file as explained in docs/changes/README.rst [2]
- revert changes in the CHANGES.rst file : git checkout master -- CHANGES.rst
- commit. If the old changelog was added in the last commit, use --amend
 
Simon

[1] https://github.com/twisted/towncrier
[2] https://github.com/astropy/astropy/blob/master/docs/changes/README.rst

**Pull request merged**